Framework/Algorithms/inc/MantidAlgorithms/DiscusMultipleScatteringCorrection.h

```diff
@@ -51,50 +51,68 @@ class MANTID_ALGORITHMS_DLL DiscusMultipleScatteringCorrection : public API::Alg
+                                                                 const size_t nXPoints, const size_t rows,
                                                                  const size_t columns);
   virtual std::unique_ptr<InterpolationOption> createInterpolateOption();
-  double interpolateFlat(std::shared_ptr<const Mantid::HistogramData::Histogram> histToInterpolate, double x);
-  double interpolateSquareRoot(const HistogramData::Histogram &histToInterpolate, double x);
+  double interpolateFlat(const API::ISpectrum &histToInterpolate, double x);
+  std::tuple<double, int> sampleQW(const API::MatrixWorkspace_sptr &CumulativeProb, double x);
+  double interpolateSquareRoot(const API::ISpectrum &histToInterpolate, double x);
+  double interpolateGaussian(const API::ISpectrum &histToInterpolate, double x);
+  double Interpolate2D(API::MatrixWorkspace_sptr SOfQ, double w, double q);
   void updateTrackDirection(Geometry::Track &track, const double cosT, const double phi);
-  void integrateCumulative(const Mantid::HistogramData::Histogram &h, double xmax, std::vector<double> &resultX,
-                           std::vector<double> &resultY);
+  void integrateCumulative(const Mantid::HistogramData::Histogram &h, double xmin, double xmax,
+                           std::vector<double> &resultX, std::vector<double> &resultY);
 
 private:
   void init() override;
   void exec() override;
   std::map<std::string, std::string> validateInputs() override;
   API::MatrixWorkspace_sptr createOutputWorkspace(const API::MatrixWorkspace &inputWS) const;
   std::tuple<double, double> new_vector(const Kernel::Material &material, double kinc, bool specialSingleScatterCalc);
-  double simulatePaths(const int nEvents, const int nScatters, Kernel::PseudoRandomNumberGenerator &rng,
-                       const HistogramData::Histogram &invPOfQ, const double kinc, const Kernel::V3D &detPos,
-                       bool specialSingleScatterCalc);
-  std::tuple<bool, double, double> scatter(const int nScatters, Kernel::PseudoRandomNumberGenerator &rng,
-                                           const HistogramData::Histogram &invPOfQ, const double kinc,
-                                           const Kernel::V3D &detPos, bool specialSingleScatterCalc);
+  std::vector<double> simulatePaths(const int nEvents, const int nScatters, Kernel::PseudoRandomNumberGenerator &rng,
+                                    API::MatrixWorkspace_sptr &invPOfQ, const double kinc,
+                                    const std::vector<double> &wValues, const Kernel::V3D &detPos,
+                                    bool specialSingleScatterCalc);
+  std::tuple<bool, std::vector<double>, double> scatter(const int nScatters, Kernel::PseudoRandomNumberGenerator &rng,
+                                                        const API::MatrixWorkspace_sptr &invPOfQ, const double kinc,
```

Comment:
There is still a variable called `kinc` in this function - it has been renamed to just `k` in other functions (presumably to reflect that ki or kf could be fixed?) - should this be renamed as well?